experimental/JuLie/partitions.jl

```diff
@@ -0,0 +1,727 @@
+    for l = 2:k
+      for m = l+1:n-l+1
+        p[m] = p[m] + p[m-l]
+      end
+    end
+    return p[n-k+1]
+  end
+
+end
+
+
+@Markdown.doc """
+    partitions(n::IntegerUnion)
+
+A list of all partitions of an integer ``n ≥ 0``, produced in
+lexicographically *descending* order. This ordering is like in Sage, but
+opposite to GAP. You can apply the function `reverse` to reverse the
+order. As usual, you may increase performance by using smaller integer types.
+The algorithm used is "Algorithm ZS1" by Zoghbi & Stojmenovic (1998); see
+[ZS98](@cite).
+
+# Examples
+```jldoctest
+julia> partitions(Int8(4))
+5-element Vector{Partition{Int8}}:
+ Int8[4]
+ Int8[3, 1]
+ Int8[2, 2]
+ Int8[2, 1, 1]
+ Int8[1, 1, 1, 1]
+```
+"""
+function partitions(n::IntegerUnion)
+
+  #Argument checking
+  n >= 0 || throw(ArgumentError("n >= 0 required"))
+
+  # Use type of n
+  T = typeof(n)
+
+  # Some trivial cases
+  if n==0
+    return Partition{T}[ Partition{T}([]) ]
+  elseif n==1
+    return Partition{T}[ Partition{T}([1]) ]
+  end
+
+  # Now, the algorithm starts
+  P = Partition{T}[]    #this will be the array of all partitions
+  k = 1
+  q = 1
+  d = fill( T(1), n )
+  d[1] = n
+  push!(P, Partition{T}(d[1:1]))
+  while q != 0
+    if d[q] == 2
+      k += 1
+      d[q] = 1
+      q -= 1
+    else
+      m = d[q] - 1
+      np = k - q + 1
+      d[q] = m
+      while np >= m
+        q += 1
+        d[q] = m
+        np = np - m
+      end
+      if np == 0
+        k = q
+      else
+        k = q + 1
+        if np > 1
+          q += 1
+          d[q] = np
+        end
+      end
+    end
+    push!(P, Partition{T}(d[1:k]))
+  end
+  return P
+
+end
+
+
+@Markdown.doc """
+    ascending_partitions(n::IntegerUnion;alg="ks")
+
+Instead of encoding a partition of an integer ``n ≥ 0`` as a *descending*
+sequence (which is our convention), one can also encode it as an *ascending*
+sequence. In the papers Kelleher & O'Sullivan (2014) and Merca (2012) it is
+said that generating the list of all ascending partitions is more efficient
+than generating descending ones. To test this, I have implemented the
+algorithms given in the papers:
+1. "ks" (*default*) is the algorithm "AccelAsc" (Algorithm 4.1) in [KO14](@cite).
+2. "m" is Algorithm 6 in [Mer12](@cite). This is actually similar to "ks".
+
+The ascending partitions are stored here as arrays and are not of type
+`Partition` since the latter are descending by our convention. I am using "ks"
+as default since it looks slicker and I believe there is a tiny mistake in the
+publication of "m" (which I fixed).
+
+# Comparison
+
+I don't see a significant speed difference to the descending encoding:
+```julia-repl
+julia> @btime partitions(Int8(90));
+3.376 s (56634200 allocations: 6.24 GiB)
+
+julia> @btime ascending_partitions(Int8(90),alg="ks");
+3.395 s (56634200 allocations: 6.24 GiB)
+
+julia> @btime ascending_partitions(Int8(90),alg="m");
+3.451 s (56634200 allocations: 6.24 GiB)
```

Comment:
Interesting! With which Julia version, OS and hardware did you obtain these benchmark results?

On my M1 MacBook with Julia 1.9.0-rc1, `ascending_partitions` is actually twice as fast as `partitions`:
```
julia> @btime partitions(Int8(90));
  8.959 s (56634195 allocations: 4.53 GiB)

julia> @btime ascending_partitions(Int8(90),alg="ks");
  4.546 s (56634195 allocations: 4.53 GiB)

julia> @btime ascending_partitions(Int8(90),alg="m");
  4.616 s (56634195 allocations: 4.53 GiB)
```

Reply:
Oh! With Julia 1.8.5, it is *much* faster and looks more like your numbers:
```julia-repl
julia> @btime partitions(Int8(90));
  2.123 s (56634195 allocations: 4.53 GiB)

julia> @btime ascending_partitions(Int8(90),alg="ks");
  2.478 s (56634195 allocations: 4.53 GiB)

julia> @btime ascending_partitions(Int8(90),alg="m");
  2.432 s (56634195 allocations: 4.53 GiB)
```
So there seems to be a major performance regression in Julia 1.9 :-(. Wonder if we can pinpoint it...

Reply:
This is weird. I believe my old benchmarks were done even with Julia 1.7.